apollo-spaceport/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<tonic::transport::Error> for ReporterError {
+    fn from(error: tonic::transport::Error) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<std::io::Error> for ReporterError {
+    fn from(error: std::io::Error) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<sys_info::Error> for ReporterError {
+    fn from(error: sys_info::Error) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<JoinError> for ReporterError {
+    fn from(error: JoinError) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl std::fmt::Display for ReporterError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "ReporterError: source: {}, message: {}",
+            self.source, self.msg
+        )
+    }
+}
+
+impl Eq for ReporterGraph {}
+
+// PartialEq is derived in the generated code, but Hash isn't and we need
+// it to use this as key in a HashMap. We have to make sure this
+// implementation always matches the derived PartialEq in the generated
+// code.
+#[allow(clippy::derive_hash_xor_eq)]
+impl Hash for ReporterGraph {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.reference.hash(state);
+        self.key.hash(state);
+    }
+}
+
+impl Report {
+    /// Try to create a new Report.
+    ///
+    /// This can fail if the ReportHeader is not created.
+    pub fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let header = Some(ReportHeader::try_new(graph)?);
+
+        Ok(Report {
+            header,
+            ..Default::default()
+        })
+    }
+}
+
+#[cfg(target_os = "windows")]
+fn get_uname() -> Result<String, std::io::Error> {
+    // Best we can do on windows right now
+    let sysname = sys_info::os_type().unwrap_or("Windows".to_string());
+    let nodename = sys_info::hostname().unwrap_or("unknown".to_string());
+    let release = sys_info::os_release().unwrap_or("unknown".to_string());
+    let version = "unknown";
+    let machine = "unknown";
+    Ok(format!(
+        "{}, {}, {}, {}, {}",
+        sysname, nodename, release, version, machine
+    ))
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_uname() -> Result<String, std::io::Error> {
+    let u = uname::uname()?;
+    Ok(format!(
+        "{}, {}, {}, {}, {},",
+        u.sysname, u.nodename, u.release, u.version, u.machine
+    ))
+}
+
+impl ReportHeader {
+    fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let mut header = ReportHeader {
+            ..Default::default()
+        };
+
+        header.hostname = hostname()?;
```

Comment:
As mentioned previously I do suspect that hostname and uname should be cached as they are possibly doing syscalls? Can this at least get a comment?

Reply:
I'm going to raise an issue. I resisted doing this earlier because I was trying to clarify my thoughts on exactly which "hostname"/uname we are going to report.

In the case where there is an external spaceport which is processing data from multiple routers (each on, potentially, different hosts) should we be reporting details about the system where the spaceport is executing or about the system where the routers are executing? (I'm fairly sure we want the latter, although currently we get the former).

At some point in the future, assuming that we do want the former, I intend to pass in a ReportHeader in with each batch of data (or have a register that associates a router with a ReportHeader in a collector or ...). Anyway, when that thinking is clarified this problem will disappear.

Reply:
tbh i think we completely ignore these two fields in Studio. Like we don't even save them as columns in Druid that we later fail to query — I think they are just thrown away without ever being read.  (It's still nice to send something reasonable in case we want to use them later or learn things, but they're not powering anything right now.)